File: src/main/java/com/EmosewaPixel/pixellib/tiles/PoweredTE.java
package com.EmosewaPixel.pixellib.tiles;

import com.EmosewaPixel.pixellib.blocks.ActivatableMachineBlock;
import com.EmosewaPixel.pixellib.recipes.EnergyMachineRecipe;
import com.EmosewaPixel.pixellib.recipes.EnergyRecipeList;
import net.minecraft.item.ItemStack;
import net.minecraft.tileentity.TileEntityType;
import net.minecraft.util.Direction;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.common.util.LazyOptional;
import net.minecraftforge.energy.CapabilityEnergy;
import net.minecraftforge.energy.IEnergyStorage;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public class PoweredTE extends AbstractRecipeBasedTE<EnergyMachineRecipe> implements IEnergyStorage {
    protected int energy = 0;
    protected int maxPower;
    protected int maxPowerIn;

    public PoweredTE(TileEntityType type, EnergyRecipeList recipeList, int maxPower, int maxPowerIn) {
        super(type, recipeList);
        setCurrentRecipe(EnergyMachineRecipe.EMPTY);
        this.maxPower = maxPower;
        this.maxPowerIn = maxPowerIn;
    }

    @Override
    public void tick() {
        if (!world.isRemote) {
            EnergyMachineRecipe recipe = getCurrentRecipe();
            if (!recipe.isEmpty()) {
                world.setBlockState(pos, world.getBlockState(pos).with(ActivatableMachineBlock.ACTIVE, true));
                if (getProgress() > 0) {
                    world.setBlockState(pos, world.getBlockState(pos).with(ActivatableMachineBlock.ACTIVE, true));
                    if (internalExtractEnergy(recipe.getEnergyPerTick(), false)) {
                        subtractProgress(1);
                        if (getProgress() == 0)
                            work();
                    }
                } else
                    startWorking();
            } else {
                world.setBlockState(pos, world.getBlockState(pos).with(ActivatableMachineBlock.ACTIVE, false));
                setProgress(0);
            }
            markDirty();
        }
    }

    public EnergyMachineRecipe getRecipeByInput() {
        Stream<ItemStack> stacksStream = IntStream.range(0, getRecipeList().getMaxInputs()).mapToObj(i -> recipeInventory.getStackInSlot(i));

        if (stacksStream.anyMatch(ItemStack::isEmpty))
            return EnergyMachineRecipe.EMPTY;

        EnergyMachineRecipe chosenRecipe = getRecipeList().getRecipes().stream().filter(recipe -> recipe.isInputValid((ItemStack[]) stacksStream.toArray())).findFirst().get();

        if (chosenRecipe == null)
            return EnergyMachineRecipe.EMPTY;

        ItemStack recipeInputs[] = (ItemStack[]) IntStream.range(0, getRecipeList().getMaxInputs()).mapToObj(i ->
                new ItemStack(recipeInventory.getStackInSlot(i).getItem(), chosenRecipe.getCountOfInputItem(recipeInventory.getStackInSlot(i)))).toArray();
        return new EnergyMachineRecipe(recipeInputs, chosenRecipe.getAllOutputs(), chosenRecipe.getTime(), chosenRecipe.getEnergyPerTick());
    }

    @Override
    public int receiveEnergy(int maxReceive, boolean simulate) {
        int energyReceived = Math.min(getMaxEnergyStored() - energy, Math.min(maxPowerIn, maxReceive));
        if (!simulate) {
            energy += energyReceived;
            markDirty();
        }
        return energyReceived;
    }

    @Override
    public int extractEnergy(int maxExtract, boolean simulate) {
        return 0;
    }

    protected boolean internalExtractEnergy(int extract, boolean simulate) {
        int energyExtract = Math.min(energy, extract);
        if (!simulate) {
            if (energyExtract != extract)
                return false;
            energy -= extract;
            markDirty();
        }
        return energyExtract == extract;
    }

    @Override
    public int getEnergyStored() {
        return energy;
    }

    @Override
    public int getMaxEnergyStored() {
        return maxPower;
    }

    @Override
    public boolean canExtract() {
        return false;
    }

    @Override
    public boolean canReceive() {
        return true;
    }

    @Nonnull
    @Override
    public <T> LazyOptional<T> getCapability(@Nonnull Capability<T> cap, @Nullable Direction side) {
        if (cap == CapabilityEnergy.ENERGY)
            return LazyOptional.of(() -> this).cast();
        return super.getCapability(cap, side);
    }
}

File: src/main/java/com/emosewapixel/pixellib/tiles/PoweredTE.kt
package com.emosewapixel.pixellib.tiles

import com.emosewapixel.pixellib.blocks.ActivatableMachineBlock
import com.emosewapixel.pixellib.recipes.EnergyMachineRecipe
import com.emosewapixel.pixellib.recipes.EnergyRecipeList
import net.minecraft.item.ItemStack
import net.minecraft.tileentity.TileEntityType
import net.minecraft.util.Direction
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.common.util.LazyOptional
import net.minecraftforge.energy.CapabilityEnergy
import net.minecraftforge.energy.IEnergyStorage

open class PoweredTE(type: TileEntityType<*>, recipeList: EnergyRecipeList, protected var maxPower: Int, protected var maxPowerIn: Int) : AbstractRecipeBasedTE<EnergyMachineRecipe>(type, recipeList), IEnergyStorage {
    protected var energy = 0

    override var currentRecipe: EnergyMachineRecipe = EnergyMachineRecipe.EMPTY

    public override val recipeByInput: EnergyMachineRecipe
        get() {
            val stacksStream = (0 until recipeList.maxInputs).map { recipeInventory.getStackInSlot(it) }

            if (stacksStream.any { it.isEmpty })
                return EnergyMachineRecipe.EMPTY

            val chosenRecipe = recipeList.recipes.first { it.isInputValid(stacksStream.toTypedArray()) }
                    ?: return EnergyMachineRecipe.EMPTY

            val recipeIndices = (0 until recipeList.maxInputs).map { chosenRecipe.getIndexOfInput(recipeInventory.getStackInSlot(it)) }

            return if (recipeIndices.contains(-1)) EnergyMachineRecipe.EMPTY else EnergyMachineRecipe(
                    (0 until recipeList.maxInputs).map { ItemStack(recipeInventory.getStackInSlot(it).item, chosenRecipe.getInputCount(recipeIndices[it])) }.toTypedArray(),
                    recipeIndices.map { chosenRecipe.getConsumeChance(it) }.toTypedArray(),
                    chosenRecipe.fluidInputs,
                    chosenRecipe.outputs,
                    chosenRecipe.outputChances,
                    chosenRecipe.fluidOutputs,
                    chosenRecipe.time,
                    chosenRecipe.energyPerTick)
        }

    override fun tick() {
        if (!world!!.isRemote) {
            val recipe = currentRecipe
            if (!recipe.isEmpty) {
                world!!.setBlockState(pos, world!!.getBlockState(pos).with(ActivatableMachineBlock.ACTIVE, true))
                if (progress > 0) {
                    world!!.setBlockState(pos, world!!.getBlockState(pos).with(ActivatableMachineBlock.ACTIVE, true))
                    if (internalExtractEnergy(recipe.energyPerTick, false)) {
                        progress--
                        if (progress == 0)
                            work()
                    }
                } else
                    startWorking()
            } else {
                world!!.setBlockState(pos, world!!.getBlockState(pos).with(ActivatableMachineBlock.ACTIVE, false))
                progress = 0
            }
            markDirty()
        }
    }

    override fun receiveEnergy(maxReceive: Int, simulate: Boolean): Int {
        val energyReceived = Math.min(maxEnergyStored - energy, Math.min(maxPowerIn, maxReceive))
        if (!simulate) {
            energy += energyReceived
            markDirty()
        }
        return energyReceived
    }

    override fun extractEnergy(maxExtract: Int, simulate: Boolean) = 0

    protected fun internalExtractEnergy(extract: Int, simulate: Boolean): Boolean {
        val energyExtract = energy.coerceAtMost(extract)
        if (!simulate) {
            if (energyExtract != extract)
                return false
            energy -= extract
            markDirty()
        }
        return energyExtract == extract
    }

    override fun getEnergyStored() = energy

    override fun getMaxEnergyStored() = maxPower

    override fun canExtract() = false

    override fun canReceive() = true

    override fun <T> getCapability(cap: Capability<T>, side: Direction?): LazyOptional<T> =
            if (cap === CapabilityEnergy.ENERGY) LazyOptional.of { this }.cast() else super.getCapability(cap, side)
}